app/models/questionnaire.rb

```diff
@@ -47,31 +89,92 @@ def all_content_ids_unique
+                                .map { |question| question[:id] }
     return if result.blank?
 
     errors.add(:content, "the following questions are missing the required :title attribute: #{result.pretty_inspect}")
   end
 
   def all_questions_have_ids
-    result = content.reject { |question| %i[raw unsubscribe].include?(question[:type]&.to_sym) }
-                    .reject { |question| question.key?(:id) }
-                    .map { |question| question[:title] }
+    result = content[:questions].reject { |question| %i[raw unsubscribe].include?(question[:type]&.to_sym) }
+                                .reject { |question| question.key?(:id) }
+                                .map { |question| question[:title] }
     return if result.blank?
 
     errors.add(:content, "the following questions are missing the required :id attribute: #{result.pretty_inspect}")
   end
 
-  def drawing_ids
-    content.select { |question| question[:type]&.to_sym == :drawing }.map { |question| question[:id] }
+  def all_ranges_have_labels
+    result = content[:questions].select { |question| %i[range].include?(question[:type]&.to_sym) }
+                                .reject { |question| non_empty_array?(question, :labels) }
+                                .map { |question| question[:id] }
+    return if result.blank?
+
+    errors.add(:content, 'the following range type questions are missing the required :labels' \
+                         " array attribute: #{result.pretty_inspect}")
+  end
+
+  def all_likert_radio_checkbox_dropdown_have_options
+    options_required_for = %i[checkbox likert radio dropdown]
+    result = content[:questions].select { |question| options_required_for.include?(question[:type]&.to_sym) }
+                                .reject { |question| non_empty_array?(question, :options) }
+                                .map { |question| question[:id] }
+    return if result.blank?
+
+    errors.add(:content, 'the following questions are missing their required :options' \
+      " array attribute: #{result.pretty_inspect}")
+  end
+
+  def non_empty_array?(question, attr)
+    question.key?(attr) && question[attr].is_a?(Array) && question[attr].size.positive?
+  end
+
+  def all_scores_have_required_atributes
+    result = content[:scores]
+             .reject { |score| score.key?(:id) && score.key?(:label) && score.key?(:ids) && score.key?(:operation) }
+             .map { |score| score[:label] || score[:id] }
+    return if result.blank?
+
+    errors.add(:content, "the following scores are missing one or more required attributes: #{result.pretty_inspect}")
+  end
+
+  def all_scores_have_nonempty_ids
+    result = content[:scores]
+             .select { |score| score[:ids].blank? }
+             .map { |score| score[:label] || score[:id] }
+    return if result.blank?
+
+    errors.add(:content, "the following scores have an empty ids attribute: #{result.pretty_inspect}")
+  end
+
+  def all_scores_use_existing_ids
```

Comment:
is there a way to move this stuff out of this class? perhaps by using concerns or something? not sure

Reply:
This says that is an example of concerns used wrong: https://medium.com/@carlescliment/about-rails-concerns-a6b2f1776d7d

Because it just moves the code and makes it less explicit.

Reply:
Because the functionality is tied to this class, it doesn't make sense to use Composition or Aggregation either.

I'll move them to a helper.

Reply:
Actually, this makes them non-private methods by default, unless we jump through a lot of hoops and declare them all private individually. I'll just leave it as is for now.